[Dependency Scanning] Consider optional dependencies of `@testable` textual dependencies with an adjacent binary module

artemcm opened the pull request:
For a `@testable` `import` in program source, if a Swift textual interface dependency is discovered, and has an adjacent binary `.swiftmodule`, only the binary module will reference the "optional" module dependencies that the built-for-testable module may need to be fully usable. 

This change, for `@testable` direct dependencies of module-under-scan, causes the scanner to open up the adjacent binary `.swiftmodule` module, and pull in its optional dependencies, and attempt to resolve them. If an optional dependency cannot be resolved on the filesystem, the scanner proceeds silently, without a diagnostic - this is a best-effort attempt to handle testable module dependencies as thoroughly as possible.

artemcm committed:
Factor out reading in Binary module dependency imports from 'SerializedModuleLoaderBase::scanModuleFile'.

artemcm committed:
[Dependency Scanning] Teach dependency scanner to resolve optional dependencies of a module

artemcm committed:
[Dependency Scanning] Pull optional dependencies from the adjacent binary module for direct  interface dependencies

For a `@Testable` import in program source, if a Swift interface dependency is discovered, and has an adjacent binary `.swiftmodule`, open up the module, and pull in its optional dependencies. If an optional dependency cannot be resolved on the filesystem, fail silently without raising a diagnostic.

artemcm committed:
[Dependency Scanning] Keep track of which direct dependencies are '@testable' for a Source dependency

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

xymus commented on include/swift/AST/ModuleDependencies.h:
```diff
@@ -109,6 +109,11 @@ class ModuleDependencyInfoStorageBase {
+  /// The set of modules which constitute optional module
+  /// dependencies for this module, such as `@_implementationOnly`
+  /// or `internal` imports.
+  std::vector<std::string> optionalModuleImports;
```
Just to clarify, a transitive `internal import` would either be ignored for a normal build or required for a `@testable import`. Only a transitive `@ _implementationOnly` is optional for a `@testable import`, or when loaded by the debugger.

artemcm replied:
Does being required for a `@testable import` mean it is also then printed into the `.swiftinterface`? Or is this a case where the `.swiftinterface` does not contain an `import` even though the binary module treats it as required?

xymus replied:
The import is not printed in the swiftinterface, but right, it's still treated as required even if it's just in the swiftmodule.

artemcm replied:
Cool, thank you for the clarification. 
Updated the PR to also query required dependencies from the adjacent binary module of a `@testable`-imported dependency. 👍🏼

artemcm committed:
[Dependency Scanning] Pull required dependencies from the adjacent binary module for direct '@testable' interface dependencies

They may be a super-set of the ones that appear in the textual interface - e.g. 'internal' imports will be contained in the adjacent binary module, but not the textual interface

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

xymus reviewed (approved):
It looks good, thanks!
